daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionVersion.scala

```diff
@@ -60,20 +52,22 @@ private[lf] object TransactionVersions
       nodes: HashMap[NodeId, UnversionedNode],
   ): VersionedTransaction[NodeId, Value.ContractId] = {
 
-    import VersionTimeline.Implicits._
-
-    val langVersions: Iterator[SpecifiedVersion] =
-      roots.iterator.map(nid => pkgLangVersions(nodes(nid).templateId.packageId))
-
-    val txVersion = assignVersions(langVersions.toList)
-
-    val versionNode: UnversionedNode => VersionedNode =
-      Node.GenNode.map3(identity, identity, VersionedValue(assignValueVersion(txVersion), _))
-    VersionedTransaction(
-      txVersion,
-      versionedNodes = nodes.transform((_, n) => Node.VersionedNode(txVersion, versionNode(n))),
-      roots = roots
-    )
+    val versionedNodes = nodes.transform { (_, node) =>
+      val nodeVersion = assignNodeVersion(pkgLangVersions(node.templateId.packageId))
+      val valueVersion = assignValueVersion(nodeVersion)
+      Node.VersionedNode(
+        nodeVersion,
+        Node.GenNode.map3(
+          identity[NodeId],
+          identity[ContractId],
+          VersionedValue[ContractId](valueVersion, _))(node),
```

Comment:
Side remark: If the node determines the version, then it will not be as easy to share values across different nodes. (This may become relevant once fat contract IDs are implemented. It's nothing to worry about for this PR or the current versioning clean-up.)

Reply:
The plan is to drop the version in the value inside the node.